### DISCUSSION TITLE
New package to load test parameters from files: parametrize_from_file

### DISCUSSION THREAD

**Discussion by kalekundert** [pytest-dev/pytest#8097] [Show and tell]
`@pytest.mark.parametrize` is very useful, but there are a few things that bug me about it:

- Long parameter lists tend to dwarf the actual test code and make everything hard to read.
- Python's syntax isn't great for data entry: a lot of braces/brackets/parentheses/quotes are required, and it's not always easy to organize data structures in a way that reflects their structure (especially when multiline strings are involved).

To address these issues, I wrote a package called [`parametrize_from_file`](https://parametrize-from-file.readthedocs.io/en/latest/).   Like the name implies, this package makes it easy to load test parameters from JSON/YAML/TOML/NestedText files.  This approach has several advantages:

- Moving the parameters to their own file keeps the actual test code concise and easy to read.
- File formats intended for data entry are easier to read and write than literal python data structures.
- Test parameters factored into files are easier to reuse, e.g. between different implementations of a library.

I've been using this package myself for a while, but I just finished cleaning it up for public consumption.  Let me know if it sounds useful!  Any comments or feedback would be appreciated.

**Comment by chrisjsewell** [comment]
Heya, just seen this, I've actually just implemented something a little similar: https://github.com/chrisjsewell/pytest-param-files

Since I wanted to centralise what I was already doing in e.g.
https://github.com/executablebooks/markdown-it-py/blob/master/tests/test_port/test_fixtures.py (how I was originally doing it),
and https://github.com/executablebooks/MyST-Parser/blob/master/tests/test_renderers/test_fixtures_docutils.py (using pytest-param-files)

So it's a little similar, but a few nice "extras" I have include; pointing to the line in the file on failures, and `--regen-file-failure` for regenerating the expected text on failures.

Happy to move to using your package though, if it could support the same features
